### PR TITLE
Issue #2541 - ConstraintGenerator updates

### PR DESCRIPTION
**Pull request by JohnTimm**
Signed-off-by: John T.E. Timm <johntimm@us.ibm.com>

### CODE REVIEW

File: conformance/fhir-ig-spl/src/test/java/com/ibm/fhir/ig/us/spl/ExamplesValidationTest.java [hunk 1]
```diff
@@ -61,9 +61,12 @@ public static boolean verify(Path p) {
             List<Issue> validate = FHIRValidator.validator().validate(r);
             for (Issue issue : validate) {
                 if (IssueSeverity.ERROR.equals(issue.getSeverity()) || IssueSeverity.ERROR.equals(issue.getSeverity())) {
-                    System.out.println(issue);
+//                  System.out.println(issue);
```
Comment (prb112) [inline conformance/fhir-ig-spl/src/test/java/com/ibm/fhir/ig/us/spl/ExamplesValidationTest.java:64]:
```suggestion
```

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java [hunk 2]
```diff
@@ -245,48 +245,59 @@ private String consequent(Node node) {
...
+        String prefix = isSlice(elementDefinition) ? discriminator(node) : prefix(node);
+        /*
+        if (isSlice(elementDefinition) ||
+                isRepeating(elementDefinition) ||
+                !isOptional(elementDefinition) ||
+                "extension".equals(getIdentifier(elementDefinition)) ||
+                hasChoiceTypeConstraint(elementDefinition)) {
+            sb.append(prefix).append(cardinality(node, prefix));
         }
+        */
```
Comment (prb112) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:259]: replaced by the above?

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java [hunk 3]
```diff
@@ -245,48 +245,59 @@ private String consequent(Node node) {
...
-        }
 
-        sb.append(cardinality(node, sb.toString()));
+            if (!isSlice(elementDefinition) || !criteriaMatches(prefix, joiner.toString())) {
+                /*
+                if (sb.length() > 0) {
+                    sb.append(" and ");
+                }
+                sb.append(prefix).append(".all(").append(joiner.toString()).append(")");
+                */
```
Comment (prb112) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:297]: replaced by the following?

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ProfileBuilder.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+            .code(Uri.of(code))
+            .profile(profile.stream()
+                .map(p -> Canonical.of(p))
+                .collect(Collectors.toList()))
+            .targetProfile(targetProfile.stream()
+                .map(p -> Canonical.of(p))
+                .collect(Collectors.toList()))
+            .build();
+    }
+}
```
Comment (prb112) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ProfileBuilder.java:285]: pretty straightforward, however it'd be great to have some degree of javadoc

File: fhir-profile/src/test/java/com/ibm/fhir/profile/test/ConstraintGeneratorTest.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+    @Test
+    public static void testConstraintGenerator12() throws Exception {
+        StructureDefinition profile = new ProfileBuilder(Observation.class, "http://ibm.com/fhir/StructureDefinition/TestObservation", "1.0.0")
+            .binding("Observation.code", binding(BindingStrength.REQUIRED, "http://ibm.com/fhir/ValueSet/vs-1", "http://ibm.com/fhir/ValueSet/max-vs-1"))
+            .build();
+        ConstraintGenerator generator = new ConstraintGenerator(profile);
+        List<Constraint> constraints = generator.generate();
+        assertEquals(constraints.size(), 1);
+//      assertEquals(constraints.get(0).expression(), "code.exists() and code.memberOf('http://ibm.com/fhir/ValueSet/vs-1', 'required') and code.memberOf('http://ibm.com/fhir/ValueSet/max-vs-1', 'required')");
```
Comment (prb112) [inline fhir-profile/src/test/java/com/ibm/fhir/profile/test/ConstraintGeneratorTest.java:205]: through the tests in this class, there are commented out elements, I assume these were the previous constraints?

**Review by prb112 — APPROVED** [review]
LGTM - left comments through it

File: conformance/fhir-ig-spl/src/test/java/com/ibm/fhir/ig/us/spl/ExamplesValidationTest.java (hunk 1 above)
Comment (JohnTimm) [reply conformance/fhir-ig-spl/src/test/java/com/ibm/fhir/ig/us/spl/ExamplesValidationTest.java:64]: Removed commented out line.

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java (hunk 2 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:259]: Removed commented out code.

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java (hunk 3 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:297]: Removed commented out code.

File: fhir-profile/src/test/java/com/ibm/fhir/profile/test/ConstraintGeneratorTest.java (hunk 5 above)
Comment (JohnTimm) [reply fhir-profile/src/test/java/com/ibm/fhir/profile/test/ConstraintGeneratorTest.java:205]: Removed commented out code.

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ProfileBuilder.java (hunk 4 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ProfileBuilder.java:481]: Added javadoc.

**Review by prb112 — APPROVED** [review]

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java [hunk 6]
```diff
@@ -389,33 +396,46 @@ private String discriminator(Node node) {
...
+                    hasDiscriminatorType(elementDefinition, DiscriminatorType.PATTERN)) {
+                joiner.add(generateVocabularyConstraint(node));
+            }
+            if (hasProfileConstraint(elementDefinition) &&
+                    hasDiscriminatorType(elementDefinition, DiscriminatorType.PROFILE)) {
+                joiner.add(generateProfileConstraint(node));
+            }
+            sb.append(joiner.toString()).append(")");
+            if (nested) {
+//              sb.append(cardinality(node, sb.toString()));
```
Comment (lmsurpre) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:436]: safe to remove?

**Review by lmsurpre — APPROVED** [review]
LGTM

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java (hunk 6 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java]: Removed commented line.

**Pull request merged**